src/main/scala/org/broadinstitute/hail/check/Gen.scala

```diff
@@ -72,6 +73,34 @@ object Gen {
     p.rng.nextUniform(min, max, true)
   }
 
+  def shuffle[T](is: IndexedSeq[T]): Gen[IndexedSeq[T]] = {
+    Gen { (p: Parameters) =>
+      if (is.isEmpty)
+        is
+      else
+        p.rng.nextPermutation(is.size, is.size).map(is)
+    }
+  }
+
+  def chooseWithWeights(weights: Array[Double]): Gen[Int] = {
+    assert(weights.forall(x => 0 <= x && x <= 1) && D_==(weights.sum, 1))
```

Comment:
This is suspiciously close to "frequency".  I'd extend frequency to have weights that are T <: Numeric[T] rather than Int.  Then you can implement:

```
  def chooseWithWeights(weights: Array[Double]): Gen[Int] =
    frequency(weights.zipWithIndex.map { case (w, i) => (w, Gen.const(i)) }: _*)
```